fix(seed-update): browser console, fCC Console, functionality of both

ed42311 opened the pull request:
<!-- freeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] Your pull request targets the `staging` branch of freeCodeCamp.
- [X] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [X] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).

- [X] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->
- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #15242 

#### Description
<!-- Describe your changes in detail -->

I've created a new challenge which goes through the process of looking at console logs and step by step encourages the user to see how each console log is processed through the tests.  Then there is a possible fix if the user would like to only console.log one variable using console.clear and finally a reference to the console web API on MDN that has many useful console.log options.

In the future might need to practice some squashing of commits...
I noticed that I spelled begining incorrectly should be beginning, but I'm guessing there might be a few more edits on this first round....

QuincyLarson commented:
@ed42311 thanks for opening this PR. 

@HKuz Since you created the Debugging section, could you take a look at this? I see a few minor issues with it (JavaScript isn't properly capitalized). I just want to make sure this challenge is as clear as can be.

Also, I think instead of sending people to the MDN, we should try to explain the concept as well as possible ourselves. Each of these challenges will have corresponding wiki articles, which they can link to from there.

HKuz commented:
Hi folks, sorry for the delayed response, you caught me switching to a new computer and in the middle of a move. I'm on spotty wifi, so likely will be early next week when I'm able to give this a proper look. Thanks for your patience, @ed42311 !

ed42311 commented:
no worries @HKuz let me know what I can do to help.  Thanks for the followup @QuincyLarson.

ed42311 commented:
Hey @QuincyLarson, this is a bit off topic, but thought it might be a decent place to start the conversation anyway.  Interest in helping with the editing on Medium for FCC, not sure you still need it since  [this post](https://medium.freecodecamp.org/seeking-a-volunteer-editor-to-help-lead-mediums-largest-technology-publication-26fe3670ea09) was awhile ago.  If you still need it, I'm down to jump in somewhere. Let me know where to jump to that conversation if it's still relevant.

QuincyLarson commented:
@HKuz OK - sounds good. We will wait for your word on this.

@ed42311 Please send an email to submit@freecodecamp.org and we can correspond regarding the volunteer Medium editing role :)

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
     },
     {
+      "id": "587d7b83367417b2b2512b37",
+      "title": "Clarifying the In-Browser Console Log",
```
I think the title should be something very obvious, like "Understand How the freeCodeCamp Console is Different than the Browser Console"

ed42311 replied:
Oh I think I see what you are referring to @HKuz. the freeCodeCamp Console in this case is the CodeMirror block that is part of the html?  Ok I think I see what you are getting at...

ed42311 replied:
On second thought maybe I am misunderstanding, going to have to run it and see...

ed42311 replied:
Was a bit confused here.  There are actually three, maybe four different things happening here, one is the explanation of the CodeMirror console and that functionality that exists inside of that, another is the explanation of the console.logs inside the CodeMirror text-editor block and how that output functions in terms of the running of the code and the print of those outputs in the Browser console.  I was only addressing one of these angles not both...

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+      "id": "587d7b83367417b2b2512b37",
+      "title": "Clarifying the In-Browser Console Log",
+      "description": [
+        "In an ordinary Javascript file loaded and run in your browser a <code>console.log()</code> will print exactly what you tell it to print to the console. In your in-browser text editor the process is slightly different and can be confusing at first. Any value passed to a <code>console.log</code> runs each set of test data as well as your function call.",
```
If someone hasn't read the issue thread, they may not understand why this challenge is here. I think we can re-work this first paragraph as follows to make it clear:

You may have noticed that some freeCodeCamp JavaScript challenges include their own console. This behaves a little differently than the browser console you used in the last challenge.

When you load and run an ordinary JavaScript file in your browser, the `console.log()` statements will print exactly what you tell it to print to the browser console. However, the freeCodeCamp console [... rework that last sentence "Any value passed..." here to clarify exactly what fCC code and console do. I also didn't understand what 'function call' was referring to.]

ed42311 replied:
Hmm... a little bit unclear what you are referring to when you say  "include their own console", it's not that the challenges include their own console, it's that the browser console behaves different based on how the code is read...  Maybe I am misunderstanding though...  I'll work this in and see what you think, if I can make git squash work this time around.  😄

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+      "description": [
+        "In an ordinary Javascript file loaded and run in your browser a <code>console.log()</code> will print exactly what you tell it to print to the console. In your in-browser text editor the process is slightly different and can be confusing at first. Any value passed to a <code>console.log</code> runs each set of test data as well as your function call.",
+        "This lends itself to some interesting behavior and might trip you up in the begining, because a line that you expect to see only once will run all of the variables presented in the tests as well as the in-browser text editor printing to the console.",
+        "If you would like to see only your single output and not have to worry about running through the test cycles you can use <code>console.clear()</code>",
```
Add a comma after "cycles" here

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "In an ordinary Javascript file loaded and run in your browser a <code>console.log()</code> will print exactly what you tell it to print to the console. In your in-browser text editor the process is slightly different and can be confusing at first. Any value passed to a <code>console.log</code> runs each set of test data as well as your function call.",
+        "This lends itself to some interesting behavior and might trip you up in the begining, because a line that you expect to see only once will run all of the variables presented in the tests as well as the in-browser text editor printing to the console.",
+        "If you would like to see only your single output and not have to worry about running through the test cycles you can use <code>console.clear()</code>",
+        "For more console based functions check out:",
```
Per @QuincyLarson comment, we should remove this line and the MDN link.

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+      "challengeSeed": [
+        "// Inspect the page with control+i or right click->inspect.",
+        "let outputTwice = \"This will print out 2 times\";",
+        "// Console log out the above variable.",
```
Say "// Use console.log() to print the outputTwice variable", then leave a couple blanks lines

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "let outputTwice = \"This will print out 2 times\";",
+        "// Console log out the above variable.",
+        "let outputFour = \"This will print to the console 4 times, twice for the logs times two tests\";",
+        "// Then console log out this ^ one...",
```
Say, "// Use console.log() to print the outputFour variable", then leave a couple blank lines

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "This lends itself to some interesting behavior and might trip you up in the begining, because a line that you expect to see only once will run all of the variables presented in the tests as well as the in-browser text editor printing to the console.",
+        "If you would like to see only your single output and not have to worry about running through the test cycles you can use <code>console.clear()</code>",
+        "For more console based functions check out:",
+        "&nbsp<a href=\"https://developer.mozilla.org/en-US/docs/Web/API/Console\">MDN Console Reference</a>"
```
Add an 'instructions' area by including a paragraph with a horizontal rule (`<hr>`), then a new paragraph that says "Use `console.log() to print the variables in the code where indicated."

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "&nbsp<a href=\"https://developer.mozilla.org/en-US/docs/Web/API/Console\">MDN Console Reference</a>"
+      ],
+      "challengeSeed": [
+        "// Inspect the page with control+i or right click->inspect.",
```
I don't think the `control-i` shortcut works across Chrome/Firefox (browsers referred to in last challenge) for all Mac/Windows/Linux users. Since the last challenge explains how to get a browser console open, we can just say "// Open your browser console" here instead of the inspect instructions.

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "// Uncomment the console.clear() and make a note of which one prints"
+      ],
+      "tests": [
+        "assert(code.match(/console\\.log\\(outputTwice\\)/g), 'message: This should print out <code>outputTwice</code> two times.');",
```
The campers won't see what's being tested, so we should adjust the test message to say, "Use `console.log()` to print the `outputTwice` variable. This should print out the value of the variable two times."

Similar comment for the second test.

Do we need that third test? It sounded like un-commenting the `console.clear()` was optional.

ed42311 replied:
Made it optional because I wasn't 100 percent sure that it applied or that it work the same way in all environments, it's something that I appreciated discovering, and allows you to log to the browser without all the random output, but it is not strictly necessary.

will update these
  - []  `'message: Use <code>console.log()</code> to print the <code>outputTwice</code> variable.  In your Browser Console this should print out the value of the variable two times.'`
  - []  `'message: Use <code>console.log()</code> to print the <code>outputFour</code> variable.  In your Browser Console this should print out the value of the variable four times.'`
-  [] remove 3rd test.  Make it a not in the text?

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "let outputFour = \"This will print to the console 4 times, twice for the logs times two tests\";",
+        "// Then console log out this ^ one...",
+        "",
+        "// Modify the following console.logs",
```
We can remove this line and the given `console.log()` statements since we're asking the camper to write them in above.

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,42 @@
+        "// console.clear();",
+        "console.log();",
+        "// Want one variable only?",
+        "// Uncomment the console.clear() and make a note of which one prints"
```
minor tweak: "...which one prints" -> "which statement prints"

HKuz reviewed (changes requested):
@ed42311 thanks for writing this new challenge - most of my comments are meant to make the challenge stand alone so campers are really clear about how fCC runs their code and what the expected behavior is for console.log() statements.

ed42311 commented:
@HKuz Awesome!  Thanks for all the edits and the clarity of thought as well, I'll take a deeper look and update the pull when work and life look away for a moment and I'm allowed some space.  :)

ed42311 committed:
fix: updating changes on debugging console from review

ed42311 commented:
kk,
 -  [x]  `'message: Use <code>console.log()</code> to print the <code>outputTwice</code> variable.  In your Browser Console this should print out the value of the variable two times.'`
  -  [x]  `'message: Use <code>console.log()</code> to print the <code>outputFour</code> variable.  In your Browser Console this should print out the value of the variable four times.'`
  -  [x] optional test, does it need to be there? Take out the test, but leave in the ability to try it.
  -  [x] minor tweak: "...which one prints" -> "which statement prints"
  -  [x] delete unneccesary code block after // Modify...
  -  [x] // Open browser console
  -  [x] OutputTwo/four specfic
  -  [x] Remove MDN link
  -  [x] "cycles, "

LMKWYT 😸

ed42311 commented:
@HKuz @QuincyLarson let me know how the changes look...

QuincyLarson commented:
@HKuz could you take a look at this when you get a moment since you're more familiar with it? If it's ready, you can merge.

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,44 @@
+        ""
+      ],
+      "tests": [
+        "assert(code.match(/console\\.log\\(outputTwice\\)/g), 'message: Use <code>console.log()</code> to print the <code>outputTwice</code> variable.  In your Browser Console this should print out the value of the variable two times.');",
```
Just fix the regex so it tests for `outputTwo` instead of `outputTwice`, and the message to update for the new variable name. Also, when I ran this locally (on Firefox), the browser console printed both statements 2x - it didn't print the `outputFour` one 4x. Is `outputFour` supposed to be in a function call?

ed42311 replied:
hmm...  maybe a difference in firefox and chrome, I'll add it to the list

HKuz replied:
@ed42311 - most of fCC's user base uses Chrome (I don't have it on my machine, though) - if the behavior is as-expected for Chrome, then leave the challenge as-is, and add a note about seeing different results based on browser. 👍

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,44 @@
+        "// Open your browser console",
+        "let outputTwo = \"This will print to the browser console 2 times\";",
+        "// Use console.log() to print the outputTwo variable",
+        "let outputFour = \"This will print to the browser console 4 times, twice for the logs times two tests\";",
```
Add 2 lines of white space under the comment `// Use console.log() to print the outputTwo variable` to give campers some space to work 👍

ed42311 replied:
switched up

HKuz commented on seed/challenges/02-javascript-algorithms-and-data-structures/debugging.json:
```diff
@@ -68,6 +68,44 @@
+        "Values passed to a <code>console.log</code> in the text editor block run each set of tests as well as one more time for any function calls that you have in you code, i.e. <code>returnSum();</code>",
+        "This lends itself to some interesting behavior and might trip you up in the beginning, because a logged value that you expect to see only once may print out many more times depending on the number of tests and the values being passed to those tests.",
+        "If you would like to see only your single output and not have to worry about running through the test cycles, you can use <code>console.clear()</code>.",
+        "The fCC console on the other hand is used for feedback while passing the test cases in fCC and does not print your <code>console.logs()</code>",
```
Just add a period at the end of this sentence.

ed42311 replied:
Added

HKuz reviewed (changes requested):
@ed42311 - this is looking good! So sorry for the delayed review, I was tied up on the Guide repo and missed these update notifications 😱 😬 . Just a few minor comments and request to check out the expected behavior for `outputFour` and we'll get this merged 👍

ed42311 commented:
No worries @HKuz.  Hope to get to this by tonight/wednesday after that it might be next week.  Going into the  mountains ⛰ for awhile.  Thanks for the edits.

ed42311 committed:
refactor(seed): Functionality for console.log test in FF and Chrom

Editing a few errors and checking functionality

BREAKING CHANGE: none that I know of

Closes #15660

ed42311 commented:
Hey @HKuz I stepped back a bit and removed the outputFour, it wasn't very clear what I was trying to do there I think, just testing for the original case of the double run through and then the option to have someone clear the console and print a single variable.  LMK what you think, I'll keep an eye on this so we can get it wrapped up.  Cheers.

HKuz commented:
Thanks, @ed42311 ! I'll take a look at this tonight or tomorrow!

HKuz committed:
Remove reference to function not used in challenge

HKuz commented:
# LGTM!! 👍 🥇 💻 

Thanks for adding this challenge (and for your patience during the review process) @ed42311 !! I just made one minor change on top of yours to remove a reference to a function that's no longer in the challenge so we could get this merged. 💥

ed42311 commented:
Ah so cool @HKuz  Thanks for all the effort in reviewing.  It's been fun.  Hope to see a few more I can work on...  Cheers.
